### PR TITLE
Added an electronic stopping fix to USER-MISC

### PR DESCRIPTION
**Pull request by jstewa-snl**
**Summary**

Added an electronic stopping fix (fix_elec_drag.*) to USER-MISC/, with corresponding examples and a README in examples/USER/misc/

**Related Issue(s)**

N/A

**Author(s)**

Both of the following are included in the fix's README file.
James Stewart (SNL), jstewa@sandia.gov
Remi Dingreville (SNL), rdingre@sandia.gov

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No issues known.

**Implementation Notes**

Downloaded master branch on 23 Sep 2020. Dropped files in USER-MISC, compiled with cmake, ran the included input files to create the included log files, which match output from lammps for the last few years we've used this fix.

**Post Submission Checklist**

- [] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

N/A

### CODE REVIEW

**Comment by akohlmey** [comment]
Thanks for your submission. Please let us know how this is different from https://docs.lammps.org/fix_electron_stopping.html ?

**Comment by jstewa-snl** [comment]
Hi, Axel. These appear to be two implementations getting at the same physics. This implementation uses a linear or quadratic fit to the stopping force (either provided by SRIM as in the other implementation, or TD-DFT simulations that don't necessarily provide as many data points for a "nice" lookup table like SRIM does) and the input is just the cutoff and polynomial coefficients for each atom species, instead of a look-up table as in the other implementation. This version also uses an additional polynomial around the low-energy cutoff to smoothly truncate the stopping power at low enough velocities so there isn't a sudden effect taking place (reflective of prior TD-DFT simulations on stopping powers).

**Comment by akohlmey** [comment]
> Hi, Axel. These appear to be two implementations getting at the same physics.

ok. looking at this from the user perspective, it would be preferable to just have one fix and then add a keyword that selects between the two variants. i also think that the name for the existing fix is more obvious than the new one and thus users will find it more easily. An alternative would be to change the names to `fix electron/stopping/xxx` and `fix electron/stopping/yyy` with xxx and yyy being a suitable keyword to tell the two apart. Then both could be documented in the same file and a few words about the differences could be added. It is probably a good idea to contact @rtoijala and discuss options. The pull request for the existing fix is #1399 so you can look for more info there.

what do you think?

I have some comments about the implementation itself, but that is something that can wait until after a solution is found to merging or renaming to get a more consistent picture for future users of this feature.

**Comment by jstewa-snl** [comment]
The latter option sounds pretty good for the time being (changing the name and adding comments to the existing docs page). I changed the name to fix_electron_stopping_fit.* in my fork to indicate that the fitted form is being used. I also recompiled with cmake and ran the example input files to make sure the electron/stopping/fit call-out works. In the meantime I can contact the authors of the other fix and see if they'd be open to a similar modification.

File: src/USER-MISC/fix_electron_stopping_fit.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+       	    gamma_z = frac*(drag_fac_2[type[i]]*v[i][2] + drag_fac_1[type[i]]);
+         } else {
+            gamma_x = drag_fac_2[type[i]]*v[i][0] + drag_fac_1[type[i]];
+            gamma_y = drag_fac_2[type[i]]*v[i][1] + drag_fac_1[type[i]];
+            gamma_z = drag_fac_2[type[i]]*v[i][2] + drag_fac_1[type[i]];
+         };
+         f[i][0] -= gamma_x*v[i][0];
+         f[i][1] -= gamma_y*v[i][1];
+         f[i][2] -= gamma_z*v[i][2];
+         f_dot_v_current += sqrt(pow(gamma_x*v[i][0], 2.0) + pow(gamma_y*v[i][1], 2.0) + pow(gamma_z*v[i][2], 2.0))*v_mag;
```
Comment (akohlmey) [inline src/USER-MISC/fix_electron_stopping_fit.cpp:178]: using the pow() function for computing a square is extremely slow. While a multiplication can be done with a single clock cycle, the pow() function requires computing a logarithm and an exponential and thus consumes hundreds of cycles and  on top of that produces a less precise result. You can use the MathSpecial::square() inline function in this case or MathSpecial::powint() for more general integer powers. This is where C/C++ differs from Fortran, which has the equivalent as built in (provided you use an integer and not a floating point number). Some modern C++ compilers provide overloaded std::pow() functions, but they are also dependent on using an integer power to access the (much) faster exponentiation. Yet for a simple square nothing can beat the square() inline function.

**Review by akohlmey — CHANGES_REQUESTED** [review]
Here are my comments. Mostly trivial and formal things. The implementation seems overall straightforward and thus not problematic or dependent on specific features.

I didn't see a .rst file for doc/src folder or modifications to the doc/src/fix_electron_stopping.rst file.

File: examples/USER/misc/elec_drag/README [hunk 2]
```diff
@@ -0,0 +1,54 @@
+// ---------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/elec_drag/README]: Folder and contents are now deleted.

File: src/USER-MISC/fix_elec_drag.cpp [hunk 3]
```diff
@@ -0,0 +1,201 @@
+/* ---------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_elec_drag.cpp]: This file is now removed.

File: src/USER-MISC/fix_elec_drag.h [hunk 4]
```diff
@@ -0,0 +1,55 @@
+/* ----------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_elec_drag.h]: This file is now removed.

File: examples/USER/misc/electron_stopping_fit/Si.tersoff.zbl [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Tersoff pair style parameters for various elements and/or mixtures
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/Si.tersoff.zbl]: Changed the in.cascade_SiSi file to read from potentials/SiC.tersoff.zbl and removed the Si.tersoff.zbl file in this folder.

File: examples/USER/misc/electron_stopping_fit/in.cascade_SiSi [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+fix              1 all nve
+fix              2 all dt/reset 1 NULL 0.001 0.05 emax 10.0
+fix              3 all electron/stopping/fit 4.63 3.3e-3 4.0e-8
+
+thermo           20
+thermo_style     custom step dt time temp pe ke f_3
+thermo_modify    lost warn flush yes
+
+dump             0 all custom 200 dump.pka_* id type x y z vx vy vz fx fy fz
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/in.cascade_SiSi]: dump and dump_modify commands are now commented out in both input files.

File: examples/USER/misc/electron_stopping_fit/log.cascase_CuNb [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    1840 7.7218826e-08 0.0012361724    18101.908   -379532.87    252702.03    287.13719 
+    1860 7.384063e-08 0.0012376841    18094.407    -379428.5    252597.32    287.48469 
+    1880 7.0826602e-08 0.0012391317    18087.172   -379327.83    252496.32    287.81735 
+    1900 6.8126364e-08 0.0012405221    18080.201   -379230.84       252399    288.13672 
+    1920 6.5698544e-08 0.0012418611    18073.494   -379137.51    252305.37    288.44417 
+    1940 6.3508807e-08 0.0012431539    18067.049   -379047.83    252215.39     288.7409 
+    1960 6.1528608e-08 0.0012444049    18060.865   -378961.79    252129.06    289.02794 
+    1980 5.9733936e-08 0.0012456182    18054.941   -378879.37    252046.37     289.3062 
+    2000 5.8104457e-08 0.0012467971    18049.277   -378800.57     251967.3    289.57652 
+Loop time of 32.2671 on 16 procs for 2000 steps with 108000 atoms
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/log.cascase_CuNb]: System size and run times in both input files have been reduced so they will run in less than 30 seconds for both 1 and 4 cpus.

File: examples/USER/misc/electron_stopping_fit/log.cascase_CuNb [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+set              atom 1 type 2
+Setting atom values ...
+  1 settings made for type
+group            pka id 1
+1 atoms in group pka
+velocity         pka set 1120 1620 400
+
+fix              1 all nve
+fix              2 all dt/reset 1 NULL 0.001 0.05 emax 10.0
+fix              3 all elec/drag 3.49 1.8e-3 9.0e-8 7.57 4.2e-3 5.0e-8
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/log.cascase_CuNb]: The log files have been deleted and will be regenerated when the source code corrections below have been implemented.

File: src/USER-MISC/fix_electron_stopping_fit.cpp [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+      drag_fac_1[i] = drag_fac_in_1[i];
+      drag_fac_2[i] = drag_fac_in_2[i];
+   };
+};
+
+// ---------------------------------------------------------------------
+
+FixElectronStoppingFit::~FixElectronStoppingFit()
+{
+   delete [] drag_fac_in_1;
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_electron_stopping_fit.cpp:107]: "delete [] energy_coh_in;" has been added.

File: src/USER-MISC/fix_electron_stopping_fit.cpp [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+   
+   scalar_flag = 1;
+   global_freq = 1;
+   
+   energy_coh_in = new double[atom->ntypes+1];
+   drag_fac_in_1 = new double[atom->ntypes+1];
+   drag_fac_in_2 = new double[atom->ntypes+1];
+   
+   for (int i = 1; i <= atom->ntypes; i++) {
+      energy_coh_in[i] = atof(arg[3*i]);
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_electron_stopping_fit.cpp]: atof() has been changed to utils::numeric().

File: src/USER-MISC/fix_electron_stopping_fit.cpp [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  " year    = { 2020 },\n"
+  " volume  = { 102 },\n"
+  " number  = { 2 },\n"
+  " pages   = { 024107 }\n"
+  "}\n\n";
+
+// ---------------------------------------------------------------------
+
+FixElectronStoppingFit::FixElectronStoppingFit(LAMMPS *lmp, int narg, char **arg) :
+ Fix(lmp,narg,arg)
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_electron_stopping_fit.cpp]: These variables have been initialized to (nullptr).

File: src/USER-MISC/fix_electron_stopping_fit.cpp (hunk 1 above)
Comment (jstewa-snl) [reply src/USER-MISC/fix_electron_stopping_fit.cpp]: pow() is now replaced with MathSpecial::square().

File: src/USER-MISC/fix_electron_stopping_fit.cpp [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+   ------------------------------------------------------------------ */
+
+/* ---------------------------------------------------------------------
+   Contributing authors: Stephen M. Foiles (SNL)
+                         James A. Stewart (SNL)
+   ------------------------------------------------------------------ */
+
+#include "fix_electron_stopping_fit.h"
+
+#include "mpi.h"
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply src/USER-MISC/fix_electron_stopping_fit.cpp]: Cleaned up this section and removed a few #include statements.

File: examples/USER/misc/electron_stopping_fit/CuNb_ZL4.eam.alloy [hunk 13]
```diff
@@ -0,0 +1,7008 @@
+#1 CuNb Potential corrected to heat of mixing in BCC and dilute FCC
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/CuNb_ZL4.eam.alloy:1]: The example input has been modified to use an existing potential and this potential file has been removed.

File: examples/USER/misc/electron_stopping_fit/log.cascase_CuNb [hunk 14]
```diff
@@ -0,0 +1,191 @@
+LAMMPS (18 Sep 2020)
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/log.cascase_CuNb]: New log files have been created and named for each case with 1 and 4 cpus.

File: examples/USER/misc/electron_stopping_fit/README [hunk 15]
```diff
@@ -0,0 +1,54 @@
+// ---------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (jstewa-snl) [reply examples/USER/misc/electron_stopping_fit/README]: All the new files have been moved into the existing electron_stopping folder and the electron_stopping_fit folder has been removed.

**Comment by jstewa-snl** [comment]
Now that these files have been incorporated into the existing electron_stopping folder, the main thing left is to appropriately edit the doc/src/fix_electron_stopping.rst file.

**Comment by jstewa-snl** [comment]
Documentation for this fix has been incorporated into the fix_electron_stopping.rst file. The original syntax block was modified to distinguish between the two fixes and a paragraph was added to the end of the description block describing this version.

**Review by akohlmey — APPROVED** [review]
I updated the code to the current master and fixed a few remaining issues with the integration into the manual. This looks good to go to me.

**Pull request merged**